crates/democracy/src/lib.rs

```diff
@@ -1091,6 +1115,41 @@ impl<T: Config> Pallet<T> {
         Ok(weight)
     }
 
+    /// determine whether or not a new referendum should be launched. This will return true
+    /// once every week.
+    fn should_launch(now: Duration) -> Result<bool, DispatchError> {
+        if now.as_secs() < NextLaunchTimestamp::<T>::get() {
+            return Ok(false);
+        }
+
+        // time to launch - calculate the date of next launch.
+
+        // convert to format used by `chrono`
+        let secs: i64 = now.as_secs().try_into().map_err(Error::<T>::from)?;
+        let now =
+            chrono::NaiveDateTime::from_timestamp_opt(secs, now.subsec_nanos()).ok_or(Error::<T>::TryIntoIntError)?;
+
+        // calculate next week boundary
+        let beginning_of_week = now.date().week(chrono::Weekday::Mon).first_day();
+        let next_week = beginning_of_week
+            .checked_add_days(Days::new(7))
```

Comment:
Would be good if the launch offset variable included this, as @nud3l pointed out we want to make it easy for downstream consumers to just so last timestamp + offset

Reply:
That doesn't make any sense - the offset constant determines the _schedule_, i.e. at which point each week the proposals get launched. Including 7 days offset in that is nonsensical. That'd be like saying that you wake up every morning at 32:00 rather than at 8:00. If you just want to fetch the concrete timestamp at which the next proposal gets launched, you can read the `NextLaunchTimestamp` storage item - no offset reading required there

Reply:
Not sure I follow your example but in my head what makes the most sense is having this recurring boundary like the launch period, but I'm not going to dispute this for the sake of brevity. Also fwiw `NextLaunchTimestamp` will be unset when we enact this but I guess that is fine since it will just launch whatever public proposal is next.

Reply:
So this is what you have implemented:
```
LaunchOffsetMillis = 9 * 60 * 60 * 1000 // 9 hours
MondayTimestamp = now() - num_days_from_monday()
NextLaunchTimestamp = MondayTimestamp + 7 + LaunchOffsetMillis
```

What I am proposing is something like this:
```
LaunchOffsetMillis = 60 * 60 * 168 * 1000 // one week
NextLaunchTimestamp = NextLaunchTimestamp + LaunchOffsetMillis
```

This assumes `NextLaunchTimestamp` is correctly set to `MondayTimestamp` as calculated in your implementation.

Reply:
To calculate the `NextLaunchTimestamp` in `n` weeks using my approach it is simply `NextLaunchTimestamp + LaunchOffsetMillis * n`, whereas in your approach it would be `MondayTimestamp + 7 * n + LaunchOffsetMillis` - where `MondayTimestamp` has to be calculated on the fly.